src/cwt/claim.rs

```diff
@@ -0,0 +1,391 @@
+use fmul_to_int::FloatMulToInt;
+use serde::{Deserialize, Serialize};
+use serde_cbor::Value;
+use time::OffsetDateTime;
+
+/// Representation of CWT claims, as defined in
+/// [RFC8392](https://datatracker.ietf.org/doc/html/rfc8392).
+pub trait Claim: Into<Value> + TryFrom<Value, Error = Error> {
+    fn key() -> Key;
+}
+
+/// Representation of the CBOR map key used to identify a claim
+/// within a CWT claims set, and restricted to text and integer values,
+/// per [RFC8392](https://datatracker.ietf.org/doc/html/rfc8392).
+pub enum Key {
+    Text(String),
+    Integer(i128),
+}
+impl From<Key> for Value {
+    fn from(key: Key) -> Value {
+        match key {
+            Key::Text(k) => Value::Text(k),
+            Key::Integer(k) => Value::Integer(k),
+        }
+    }
+}
+
+/// Errors that can occur when parsing values into claims.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("claim requires NumericDate (int or float)")]
+    NumericDateRequired,
+    #[error("claim requires Bytes value")]
+    BytesValueRequired,
+    #[error("conversion error: {0}")]
+    ConversionError(#[from] serde_cbor::Error),
+    #[error("TryFromIntError: {0}")]
+    TryFromIntError(#[from] std::num::TryFromIntError),
+    #[error("OffsetComponentError: {0}")]
+    OffsetComponentError(#[from] time::error::ComponentRange),
+    #[error("NumericDate conversion overflows")]
+    ConversionOverflow,
+}
+
+/// Numerical representation of seconds relative to the Unix Epoch,
+/// as defined in [RFC7049](https://www.rfc-editor.org/rfc/rfc7049#section-2.4.1)
+/// with the leading tag 1 omitted.
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub enum NumericDate {
```

Comment:
I believe you need to do this otherwise it will serialize as a map:
https://serde.rs/enum-representations.html
```suggestion
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(untagged)]
pub enum NumericDate {
```

Reply:
Ahh I looked into that but it unfortunately deserializes ints into Fractional dates as well -- do you know of a way around this other than writing an entire Deserializer implementation for this? I think this shouldn't affect the claimsSet serialization since we convert to `Value` before serializing anyways, but wdyt?

Reply:
Wow, that's unexpected! You can use `#[serde(try_from="Value", into="Value")]` instead of `#[serde(untagged)]`